xsd, docs, and test for sort_by (discover_datasets)

bernt-matthias opened the pull request:
Since `planemo lint` failed so far for this attribute we can be quite sure that this was never used (in a tool)

- order of sort_comp and sort_by was wrong for sort_by in collections
- sort_by was missing for data output
- adds a test (though we can't check the generated order .. I manually tested with `planemo serve`)

follow up on https://github.com/galaxyproject/galaxy/pull/9674

bernt-matthias committed:
fix docs & add test for discover_datasets sort_by

- order of sort_comp and sort_by was wrong for
  sort_by in collections
- sort_by was missing for data output
- adds a test (though we can't check the generated
  order .. I manually tested with `planemo serve`)

mvdbeek commented on test/functional/tools/discover_sort_by.xml:
```diff
@@ -0,0 +1,26 @@
+<tool id="collection_split_on_column" name="collection_split_on_column" version="0.1.0">
```
You have to add the tool to test/functional/tools/samples_tool_conf.xml to run the tests

bernt-matthias replied:
Hah. Just forgot to commit it.

mvdbeek commented on test/functional/tools/discover_sort_by.xml:
```diff
@@ -0,0 +1,26 @@
+  <tests>
+    <test>
+      <param name="input1" value="tinywga.fam" />
+      <output_collection name="collection_numeric_name" type="list" count="11"/>
```
Don't you want to ascertain the new sort order ? You can list `<element/>` items here

bernt-matthias replied:
That's true, but I did not expect the order of the elements is relevant. Lets try.

bernt-matthias replied:
Test currently do not check for the order, but only if the elements contained in the test definition are produced by the test: https://github.com/galaxyproject/galaxy/blob/94390c2af8fb1ce8f54c6665810991da36e6331a/lib/galaxy/tool_util/verify/interactor.py#L717

This would be (quite) easy to change, but 

- the test definition is a dict and dictionary order can only be used for python >= 3.7 
- dictionary order would always be lexicographic as defined for strings
- it seems that the dictionary order is lost in the `get_tool_tests` function called here https://github.com/galaxyproject/galaxy/blob/94390c2af8fb1ce8f54c6665810991da36e6331a/lib/galaxy/tool_util/verify/interactor.py#L803

Seems difficult. 

Spontaneous ideas

- sort the test elements in verify_elements as given by `sort_by`
- leave it as is and consider the test just checking for bugs

mvdbeek replied:
> * the test definition is a dict and dictionary order can only be used for python >= 3.7

3.6, it was made official in 3.7



> * dictionary order would always be lexicographic as defined for strings

this should be by insertion order on newer python's, right ?

Anyway, I agree this is not as straightforward. Maybe add a <!--TODO: test actual order --> in the test, and then we can reconsider once sort order is guaranteed or we decide to write an API test, which could also assert this.

bernt-matthias replied:
> 3.6, it was made official in 3.7

true

> this should be by insertion order on newer python's, right ?

also true, my mistake

> Anyway, I agree this is not as straightforward. Maybe add a in the test, and then we can reconsider once sort order is guaranteed or we decide to write an API test, which could also assert this.

Still I have the urge to understand this :) Somewehere here:

https://github.com/galaxyproject/galaxy/blob/94390c2af8fb1ce8f54c6665810991da36e6331a/lib/galaxy/tool_util/verify/interactor.py#L141

the order seems to get lost. Any suggestions where I could dig?

bernt-matthias committed:
fix sorting of coutput collection elements in test definition

mvdbeek commented on lib/galaxy/web/framework/decorators.py:
```diff
@@ -326,7 +326,7 @@ def format_return_as_json(rval, jsonp_callback=None, pretty=False):
     Use `pretty=True` to return pretty printed json.
     """
-    dumps_kwargs = dict(indent=4, sort_keys=True) if pretty else {}
+    dumps_kwargs = dict(indent=4) if pretty else {}
```
I don't think we should do this until we sunset python 3.5. You could add the index of the element [here](https://github.com/mvdbeek/galaxy/blob/dev/lib/galaxy/tool_util/parser/xml.py#L601) and then sort the test elements where needed. Do you maybe want to just add the xsd changes in for 20.05 ?

bernt-matthias replied:
Your are probably right -- even if it seems safe in most places where `format_return_as_json` is used (pretty is either set to `False` or set to `True` only in debug mode). 

The only place where its `True` is here: 

https://github.com/galaxyproject/galaxy/blob/7eb2fcb708e7b63e17800c87613ddfa5497c0654/lib/galaxy/webapps/galaxy/api/workflows.py#L485

An alternative might be to add another parameter (e.g. `sort_keys=False`) to `format_return_as_json`. Tell me your preference and I will go ahead :)

mvdbeek replied:
I'd go with the xsd changes only, but if you want to dig deeper I think that adding the index of the element to the test definition seems more solid to me.

bernt-matthias committed:
py<3.6 compatible check collection element sorting

by adding an element_index attribute to the element
attributes.

background: test collection elements were sorted
alphabetically because:

1 they are stored in a dict which is unsorted in
  py<3.6
2 galaxy.web.framework.format_return_as_json
  dumps this dict to json with sort_keys=True
  (at least in debug mode which affects planemo t)

dropping 2 would be sufficient for py>=3.6, since
the test collection elements would be stored in
insertion order in the dict.

mvdbeek commented on test/functional/tools/discover_sort_by.xml:
```diff
@@ -0,0 +1,77 @@
+         - test is only if the elements listed in the test
+           are in the list of generated elements
+         - furthermore the order of the elements listed in the test
+           is not preserved in the test functions -->
```
I'm not quite sure I understand the comment, but also, does the comment still apply ?

bernt-matthias replied:
you are right. I removed it.

mvdbeek commented on lib/galaxy/tool_util/verify/interactor.py:
```diff
@@ -715,16 +709,35 @@ def get_element(elements, id):
-                template = "Failed to find identifier [%s] for testing, tool generated collection elements [%s]"
-                message = template % (element_identifier, element_objects)
+            if element is None:
+                template = "Failed to find identifier '%s' of test collection %s in the tool generated collection elements %s (at the correct position)"
```
~~Isn't `the corrent position` `i` ?~~
Is the correct position the index of the iteration over sorted_test_ids ?

mvdbeek replied:
and wouldn't this mean you can simplify the logic to

```python
for element_index, expected_element_identifier in enumerate(sorted_test_ids):
    assert element_objects[element_index]["element_identifier"] == expected_element_identifier
```
?

(You could add some length handling for missing elements)

bernt-matthias replied:
No, the test is just checking if the elements listed in the test occur in the same order in the generated elements. 

Thereby the tool author does not need to list all elements (which might be very many).

mvdbeek replied:
Alright, then how about this:
```python
for element_index, expected_element_identifier in enumerate(sorted_test_ids):
    if expected_element_identifier is None:
        break
    assert element_objects[element_index]["element_identifier"] == expected_element_identifier
```
That way it's clear at which position the test data differs, and I think the logic is a bit easier to follow ?

bernt-matthias replied:
I don't see why this should work :( .. 

The `None` entries from here: https://github.com/galaxyproject/galaxy/blob/a5b898c35bee08ada119bd0f7c3912242b1bae1c/lib/galaxy/tool_util/verify/interactor.py#L728

should all be overwritten by the actual identifiers. I only use `[None] * len(element_tests)` to initialize a list of the needed length.

I extended the test in order to clarify what I meant. The tool new generates 10 elements, but the test only uses three of them to verify the sorting order. The number of generated elements is tested with the `count` attribute. 

Formally spoken we test if the list of elements in the test is a subsequence of the generated elements.

mvdbeek replied:
what is element_tests ? Is that not the list of specified tests?

bernt-matthias replied:
Its a `dict`, looking like this

```
{'1': [None, {'assert_list': [{'attributes': {'expression': '^.*$'}, 'children': [], 'tag': 'has_text_matching'}], 'checksum': None, 'compare': 'diff', 'decompress': False, 'delta': 10000, 'delta_frac': None, 'element_index': 0, 'elements': {}, 'extra_files': [], 'lines_diff': 0, 'md5': None, 'metadata': {}, 'primary_datasets': {}, 'sort': False}], '10': [None, {'assert_list': [{'attributes': {'expression': '^.*$'}, 'children': [], 'tag': 'has_text_matching'}], 'checksum': None, 'compare': 'diff', 'decompress': False, 'delta': 10000, 'delta_frac': None, 'element_index': 2, 'elements': {}, 'extra_files': [], 'lines_diff': 0, 'md5': None, 'metadata': {}, 'primary_datasets': {}, 'sort': False}], '2': [None, {'assert_list': [{'attributes': {'expression': '^.*$'}, 'children': [], 'tag': 'has_text_matching'}], 'checksum': None, 'compare': 'diff', 'decompress': False, 'delta': 10000, 'delta_frac': None, 'element_index': 1, 'elements': {}, 'extra_files': [], 'lines_diff': 0, 'md5': None, 'metadata': {}, 'primary_datasets': {}, 'sort': False}]}
```

jmchilton replied:
Maybe calling it element_index is a bad idea if it isn't expected to match the actual element index. Something like expected_order_index maybe would be better?

I don't really have the throughput to deal with this, but maybe this change requiring the order to match should be hidden behind a profile version tag for the tool? If there wasn't a ton of tool breakage maybe it is not worth the effort, I will just update Planemo regardless.

bernt-matthias committed:
remove outdated comment

bernt-matthias committed:
clarify test

create more elements, but still list only 3 in the test

jmchilton commented on lib/galaxy/tool_util/parser/xml.py:
```diff
@@ -605,12 +605,13 @@ def __parse_output_collection_elem(output_collection_elem):
         if identifier is None:
             raise Exception("Test primary dataset does not have a 'identifier'")
         element_tests[identifier] = __parse_test_attributes(element, element_attrib, parse_elements=True)
+        element_tests[identifier][1]["element_index"] = idx
```
This breaks YAML parsing because the general interface assumes this is here but only the XML parser adds it. This would break CWL tools, YAML tools, and Galaxy workflows (this last one breaks Planemo CI after this was published). So I'm going to refactor this addition into TestCollectionOutputDef - it will just make sure this is populated there. I hope that makes sense.
